pkg/zcapld/zcap.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package zcapld
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/hyperledger/aries-framework-go/pkg/doc/verifiable"
+)
+
+const (
+	proofPurposeField         = "proofPurpose"
+	proofCapabilityChainField = "capabilityChain"
+)
+
+// Capability is a ZCAP.
+type Capability struct {
+	ID               string             `json:"id"`
+	Invoker          string             `json:"invoker"`
+	Controller       string             `json:"controller"`
+	Delegator        string             `json:"delegator"`
+	Parent           string             `json:"parentCapability"`
+	Nonce            string             `json:"nonce"`
+	AllowedAction    []string           `json:"allowedAction"`
+	InvocationTarget InvocationTarget   `json:"invocationTarget"`
+	Proof            []verifiable.Proof `json:"proof"`
+}
+
+// InvocationTarget is the target on which the capability applies.
+type InvocationTarget struct {
+	ID   string
+	Type string
+}
+
+// Invokers are this capability's entities authorized to invoke the invocation target.
+func (c *Capability) Invokers() ([]string, error) {
+	// if neither an invoker, controller, nor id is found on the capability then
+	// the capability can not be invoked
+	if c.Invoker == "" && c.Controller == "" && c.ID == "" {
+		return nil, errors.New("invoker not found for capability")
+	}
+
+	// if there's a delegator present and not an invoker, then this capability
+	// was intentionally meant to not be invoked
+	if c.Delegator != "" && c.Invoker == "" {
+		return nil, nil
```

Comment:
return non nil error here

Reply:
The reference implementation does not return an error for this condition.

Reply:
this breaks our return coding convention which return either a non nil value or non nil error, not both

the caller can deal with the return the way they want, we don't wanna break the convention.

Reply:
https://github.com/digitalbazaar/ocapld.js/blob/8a54398162837b1cf52c82978bc8127e52d02974/lib/utils.js#L52-L56

Reply:
@Baha-sk 

> this breaks our return coding convention which requires the return of either a non nil value or non nil error, not both nil.
> 
> the caller can deal with the return the way they want, we don't wanna break this convention.

Sorry, I hadn't seen this comment.

I disagreed with that convention then and still do today... but how about if I just un-export this method instead? I don't anticipate external consumers any time soon. Can be easily wrapped in a function that returns an error if we still wish so.

Reply:
@llorllale that's a JS detail implementation, which breaks our Go coding convention

I personally prefer not breaking our convention and probably ignore a non nil error at the consumer side of the call.

@troyronda @fqutishat @sandrask @bstasyszyn @rolsonquadras @sudeshrshetty, opinion?

Reply:
@Baha-sk nil is ok here since it is a slice. it will not affect the client/caller. I agree with `coding convention which requires the return of either a non nil value or non nil error` but not in that particular case. Returning non-empty slice here does not make sense and it would allocate memory. This behavior is ok.

Reply:
I un-exported call methods on `Capability` as they're not required to be exposed.

Reply:
@soluchok if the expected result is an empty slice in this case, then return an empty slice at least, not nil

again, this is not about arguing opinions, we want code to be easily maintainable and behave consistently across the board, this is why we have conventions.

Reply:
@Baha-sk the code is now returning empty slices (exactly the same as `nil`, that's how Go works) and `nil` error.

Does that clear things up for you?

Reply:
yes, it's more consistent this way @llorllale